providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelConnector.java

```diff
@@ -167,19 +156,40 @@ private synchronized void onConnectionLost() {
     }
 
     /**
-     * Restarts the event stream using the asynchronous service stub, applying a deadline if configured.
-     * Emits a connection event if the restart is successful.
-     */
-    private synchronized void restartStream() {
-        if (connected) {
-            log.debug("(Re)initializing event stream.");
-            T localServiceStub = this.serviceStub;
-            if (streamDeadlineMs > 0) {
-                localServiceStub = localServiceStub.withDeadlineAfter(this.streamDeadlineMs, TimeUnit.MILLISECONDS);
+     * Monitors the state of a gRPC channel and triggers the specified callbacks based on state changes.
+     *
+     * @param expectedState     the initial state to monitor.
+     * @param channel           the ManagedChannel to monitor.
+     * @param onConnectionReady callback invoked when the channel transitions to a READY state.
+     * @param onConnectionLost  callback invoked when the channel transitions to a FAILURE or SHUTDOWN state.
+     */
+    public static void monitorChannelState(
+            ConnectivityState expectedState,
+            ManagedChannel channel,
+            Runnable onConnectionReady,
+            Runnable onConnectionLost) {
+        channel.notifyWhenStateChanged(expectedState, () -> {
+            ConnectivityState currentState = channel.getState(true);
+            log.debug("Channel state changed to: {}", currentState);
+            if (currentState == ConnectivityState.READY) {
```

Comment:
currently we do have `waitForReady` which wlll force communication on an idle channel. But should we maybe also add here the `IDLE` state? - currently i think this path is not used anyways